components/accelerator/nvidia/nccl/component.go

```diff
@@ -4,59 +4,92 @@ package nccl
 
 import (
 	"context"
+	"fmt"
+	"os"
+	"runtime"
+	"sync"
 	"time"
 
 	apiv1 "github.com/leptonai/gpud/api/v1"
 	"github.com/leptonai/gpud/components"
 	"github.com/leptonai/gpud/pkg/eventstore"
 	"github.com/leptonai/gpud/pkg/kmsg"
 	"github.com/leptonai/gpud/pkg/log"
+	nvidianvml "github.com/leptonai/gpud/pkg/nvidia-query/nvml"
 )
 
 const Name = "accelerator-nvidia-nccl"
 
 var _ components.Component = &component{}
 
 type component struct {
-	kmsgSyncer  *kmsg.Syncer
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	nvmlInstance nvidianvml.InstanceV2
+
 	eventBucket eventstore.Bucket
+	kmsgSyncer  *kmsg.Syncer
+
+	readAllKmsg func(context.Context) ([]kmsg.Message, error)
+
+	lastMu   sync.RWMutex
+	lastData *Data
 }
 
-func New(ctx context.Context, eventStore eventstore.Store) (components.Component, error) {
-	eventBucket, err := eventStore.Bucket(Name)
-	if err != nil {
-		return nil, err
+func New(gpudInstance *components.GPUdInstance) (components.Component, error) {
+	cctx, ccancel := context.WithCancel(gpudInstance.RootCtx)
+	c := &component{
+		ctx:          cctx,
+		cancel:       ccancel,
+		nvmlInstance: gpudInstance.NVMLInstance,
 	}
 
-	kmsgSyncer, err := kmsg.NewSyncer(ctx, Match, eventBucket)
-	if err != nil {
-		return nil, err
+	if gpudInstance.EventStore != nil && runtime.GOOS == "linux" {
+		var err error
+		c.eventBucket, err = gpudInstance.EventStore.Bucket(Name)
+		if err != nil {
+			ccancel()
+			return nil, err
+		}
+
+		if os.Geteuid() == 0 {
+			c.kmsgSyncer, err = kmsg.NewSyncer(cctx, Match, c.eventBucket)
+			if err != nil {
+				ccancel()
+				return nil, err
+			}
+		}
 	}
 
-	return &component{
-		kmsgSyncer:  kmsgSyncer,
-		eventBucket: eventBucket,
-	}, nil
+	if runtime.GOOS == "linux" && os.Geteuid() == 0 {
+		c.readAllKmsg = kmsg.ReadAll
+	}
+
+	return c, nil
 }
 
 func (c *component) Name() string { return Name }
 
-func (c *component) Start() error { return nil }
+func (c *component) Start() error {
+	// do not need periodic kmsg checks since it already has a watcher
+	return nil
+}
 
-func (c *component) HealthStates(ctx context.Context) (apiv1.HealthStates, error) {
-	return []apiv1.HealthState{
+func (c *component) LastHealthStates() apiv1.HealthStates {
+	return apiv1.HealthStates{
```

Comment:
Do we need to return an empty state?  And there is no Name in the state, did we miss it?

Reply:
oh good catch, fixing